plugins/auth-backend/src/lib/passport/BackstageIdentityStrategy.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Request } from 'express';
+import { JWK, JWT, JWKS } from 'jose';
+import { BackstageIdentity } from '../../providers';
+import { IdentityClient } from '../../identity';
+import { PluginEndpointDiscovery } from '@backstage/backend-common';
+
+const Strategy = require('passport-strategy');
+// Using singleton keyStore instead of membership due to Passport wtf
+let keyStore: JWKS.KeyStore;
+let keyStoreUpdated: number;
+
+export class BackstageIdentityStrategy extends Strategy {
+  private readonly client: IdentityClient;
+  private readonly discovery: PluginEndpointDiscovery;
+
+  constructor(options: { discovery: PluginEndpointDiscovery }) {
+    super();
+    this.client = new IdentityClient({ discovery: options.discovery });
+    this.discovery = options.discovery;
+    this.name = 'backstage';
+  }
+
+  private async refreshKeyStore(rawJwtToken: string) {
+    const { header, payload } = JWT.decode(rawJwtToken, {
+      complete: true,
+    }) as {
+      header: { kid: string };
+      payload: { iat: number };
+    };
+    // Refresh public keys from identity if needed
+    if (
+      !keyStore ||
+      (!keyStore.get({ kid: header.kid }) &&
+        payload?.iat &&
+        payload.iat > keyStoreUpdated)
+    ) {
+      const now = Date.now() / 1000;
+      const publicKeys = await this.client.listPublicKeys();
+      keyStore = new JWKS.KeyStore(publicKeys.keys.map(key => JWK.asKey(key)));
+      keyStoreUpdated = now;
+    }
+  }
+
+  async authenticate(req: Request) {
+    if (
+      !req.headers.authorization ||
+      !req.headers.authorization.startsWith('Bearer ')
+    ) {
+      this.fail(new Error('No bearer token found in authorization header'));
+    }
+
+    try {
+      const token = req!.headers!.authorization!.substring(7);
+      const issuer = await this.discovery.getExternalBaseUrl('auth');
+      await this.refreshKeyStore(token);
+      const decoded = JWT.IdToken.verify(token, keyStore, {
```

Comment:
This makes an implicit assumption that the algorithm uses `ES256` to sign the keys, which could be incorrect, unless the Token was issued by Backstage itself.

Reply:
Indeed. I have two thoughts on this:
* If the token was not issued by Backstage, I think it is easier for the adopter to configure an existing passport strategy, such as http://www.passportjs.org/packages/passport-jwt/. The `BackstageIdentityStrategy` is supplied for convenience to support the default identity tokens supplied out of the box.
* Currently verification and issuance (is that the right word?) of backstage tokens are done in two different places, duplicating choice of algorithm and claims. It is likely wiser to configure that in one place to avoid unnecessary bugs when these diverge. I just didn't know a nice way to do it and thought it might be a later refactoring.

Reply:
I think we can assume that this would only be used to verify tokens that are issued by Backstage.

Either way the signing algorithm is communicated via `alg` in the JWT header iirc, so it would just be a matter of listing the allowed algorithms here. If needed we can make it configurable but sticking to just `ES256` should be good for now.